src/server.py

```diff
@@ -1,45 +1,117 @@
 # -*- coding:utf-8 -*-
 from __future__ import unicode_literals
 import socket
+import email.utils
+
+
+class HTTPException(BaseException):
+    """Custom exception class."""
+
+    def __init__(self, code, reason):
+        self.http_error = code
+        self.message = reason
+
+
+HTTP_BAD_REQUEST = '400 Bad Request'
 
 
 def server():
     """Start running HTTP server."""
     server = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_TCP)
-    address = ('127.0.0.1', 5000)
+    address = ('127.0.0.1', 5001)
     server.bind(address)
     server.listen(1)
     print('starting')
     try:
         while True:
             conn, addr = server.accept()
-            buffer_length = 8
-            message_complete = False
-            full_string = b""
-            while not message_complete:
-                part = conn.recv(buffer_length)
-                full_string = full_string + part
-                if len(part) < buffer_length:
-                    message_complete = True
-            print(full_string.decode('utf8'))
-            full_string += b"\r\n\r\n" 
-            conn.sendall(full_string)
-            conn.sendall(response_ok())
+            try:
+                buffer_length = 8
+                message_complete = False
+                full_string = b""
+                while not message_complete:
+                    part = conn.recv(buffer_length)
+                    full_string = full_string + part
+                    if len(part) < buffer_length:
+                        full_string = full_string.decode('utf-8')
+                        message_complete = True
+                print('full_string.decode', full_string)
+                try:
+                    uri = parse_request(full_string)
+                    response = (response_ok(uri))
+                except SyntaxError:
+                    response = (response_error('404 Page Not Found'))
+                except NameError:
+                    response = (response_error('405 Method Not Allowed'))
+                except TypeError:
+                    response = (response_error('505 HTTP Version Not Supported'))
+                except ValueError:
+                    raise HTTPException(
+                        HTTP_BAD_REQUEST,
+                        "No 'Host :' in header")
+            except SystemError:
+                response = (response_error())
```

Comment:
what are you expecting to cause a `SystemError`?  Or is this intended to be a "catch-all"?  If the latter, use a bare except:

```python
try:
    something_that_might_fail()
except AnExpectedException:
    handle_it()
except:
    keep_going_no_matter_what()
```